src/compute/src/render/persist_source.rs

```diff
@@ -0,0 +1,209 @@
+use timely::dataflow::{Scope, Stream};
+use timely::progress::Antichain;
+
+use mz_dataflow_types::{DataflowError, DecodeError, SourceError, SourceErrorDetails};
+use mz_expr::SourceInstanceId;
+use mz_persist::location::ExternalError;
+use mz_persist_client::read::ListenEvent;
+use mz_persist_client::Location;
+use mz_repr::{Diff, Row, Timestamp};
+
+use crate::compute_state::ComputeState;
+
+/// Creates a new source that reads from a persist shard.
+pub fn persist_source<G>(
+    _render_state: &mut ComputeState,
+    scope: &G,
+    instance_id: SourceInstanceId,
+    consensus_uri: String,
+    blob_uri: String,
+    shard_id: mz_persist_client::ShardId,
+    as_of: Antichain<Timestamp>,
+) -> (
+    Stream<G, (Row, Timestamp, Diff)>,
+    Stream<G, (DataflowError, Timestamp, Diff)>,
+    Rc<dyn Any>,
+)
+where
+    G: Scope<Timestamp = mz_repr::Timestamp>,
+{
+    let worker_index = scope.index();
+
+    let async_stream = async_stream::try_stream! {
+        // We are reading only from worker 0. We can split the work of reading from the snapshot to
+        // multiple workers, but someone has to distribute the splits. Also, in the glorious
+        // STORAGE future, we will use multiple persist shards to back a STORAGE collection. Then,
+        // we can read in parallel, by distributing shard reading amongst workers.
+        if worker_index != 0 {
+            log::trace!("We are not worker 0, exiting...");
+            return;
+        }
+
+        let timeout = Duration::from_secs(60);
+
+        let persist_location = Location {
+            consensus_uri: consensus_uri,
+            blob_uri: blob_uri,
+        };
+
+        let (blob, consensus) = persist_location.open(timeout).await?;
+
+        let persist_client =
+            mz_persist_client::Client::new(timeout, blob, consensus).await?;
+
+
+        let (_write, read) =
+            persist_client.open::<Row, Row, Timestamp, Diff>(timeout, shard_id).await?;
+
+            let mut splits = read
+                .snapshot(timeout, as_of.clone(), NonZeroUsize::new(1).unwrap())
+                .await?
+                .expect("invalid usage");
+
+            assert_eq!(splits.len(), 1);
+            let split = splits.pop().unwrap();
+            let mut iter = read
+                .snapshot_iter(timeout, split)
+                .await?
+                .expect("invalid usage");
+
+            loop {
+                let next = iter.poll_next(timeout).await?;
+                if next.is_empty() {
+                    break;
+                }
+                yield ListenEvent::Updates(next);
+            }
+
+            let mut listen = read
+                .listen(timeout, as_of)
+                .await?
+                .expect("invalid usage");
+
+            loop {
+                let next = listen.poll_next(timeout).await?;
+                for event in next {
+                    yield event;
+                }
+            }
```

Comment:
I can't really tell what any of this does. That may be fine if everyone else gets it, but I bet some comments would help generally!

Reply:
I'll add more comments explaining this.